src/components/dataentry/Dhis2Formula.js

```diff
@@ -0,0 +1,107 @@
+import React, { useEffect, useState, useContext } from "react";
+import { TextField, Tooltip } from "@material-ui/core";
+import FormDataContext from "./FormDataContext";
+
+const parseDependencies = (expression) => {
+  const components = expression
+    .split("+")
+    .flatMap((s) => s.split(","))
+    .flatMap((s) => s.split("-"))
+    .flatMap((s) => s.split("*"))
+    .flatMap((s) => s.split("ROUND("))
+    .flatMap((s) => s.split("IF("))
+    .flatMap((s) => s.split("SAFE_DIV("))
+    .flatMap((s) => s.split("ABS("))
+    .flatMap((s) => s.split("("))
+    .flatMap((s) => s.split(")"))
+    .map((s) => s.trim());
+  const deReferences = components.filter((s) => s.includes("#{"));
+  const deps = deReferences.map((de) => de.replace("#{", "").replace("}", ""));
+  return Array.from(new Set(deps));
+};
+
+const safeDiv = (a, b) => {
+  if (b !== 0) {
+    return a / b;
+  }
+  return 0;
+};
+
+const abs = (a) => {
+  return Math.abs(a);
+};
+
+const iff = (a, b, c) => {
+  return a ? b : c;
+};
+const roundFunction = (a,position)=> {
+    return a.toFixed(position)
+}
+
+const Dhis2Formula = ({ formula }) => {
+  const formDataContext = useContext(FormDataContext);
+  const [rawValue, setRawValue] = useState("");
+  const [expression, setExpression] = useState("");
+  const [error, setError] = useState(undefined);
+  useEffect(() => {
+    if (formDataContext && formula) {
+      const dependencies = parseDependencies(formula);
+      let expression = formula;
+      dependencies.forEach((dep) => {
+        const val = formDataContext.getValue(dep);
+        const valExpression = val !== undefined ? val.value : 0;
+        expression = expression.split("#{" + dep + "}").join(" " + valExpression + " ");
+      });
+
+      setExpression(expression);
+      try {
+        const evaluatedExpression = new Function(
```

Comment:
Veru cool stuff.

Reply:
this is way cooler : hesabu engine without dependency/parsing (need to replace a by this.a()) https://gist.github.com/mestachs/c848348dbe72ccb17c8896991ed0186e